### PR TITLE
Add TFBartForSequenceClassification

### PR DESCRIPTION
**Pull request by uglyboxer**
# What does this PR do?
This adds a sequence classification head to the TensorFlow implementation of BART, following the pattern of `BartForSequenceClassification` (PyTorch version)

Fixes # [Issue 19653](https://github.com/huggingface/transformers/issues/19653)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten, @patil-suraj

### CODE REVIEW

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 1]
```diff
@@ -1465,3 +1484,142 @@ def _reorder_cache(past, beam_idx):
...
+
+        masked = tf.reshape(
+            tf.boolean_mask(last_hidden_state, eos_mask),
+            (tf.shape(input_ids)[0], tf.shape(self_masked)[1], tf.shape(last_hidden_state)[-1]),
+        )
+
+        sentence_representation = masked[:, -1, :]
+        logits = self.classification_head(sentence_representation)
+        reshaped_logits = tf.reshape(tensor=logits, shape=(-1, self.config.num_labels))
+        loss = None if labels is None else self.hf_compute_loss(labels=labels, logits=reshaped_logits)
```
Comment (ydshieh) [inline src/transformers/models/bart/modeling_tf_bart.py:1589]: I feel the reshape of logits here is not necessary. Could you check again?

**Review by ydshieh — APPROVED** [review]
Hi @uglyboxer Thank you for adding this. LGTM, but I left a few comments.

File: tests/test_modeling_tf_common.py [hunk 2]
```diff
@@ -171,6 +170,12 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False) -> d
                 for k, v in inputs_dict.items()
             }
 
+        if model_class.__name__ == "TFBartForSequenceClassification":
```
Comment (ydshieh) [inline tests/test_modeling_tf_common.py:173]: I think we are able (and it's better) to remove this one, and taking care of the eos token ids in TF bart model testing file. (As I didn't see the same block in `test_modeling_common.py`)

File: tests/models/bart/test_modeling_tf_bart.py [hunk 3]
```diff
@@ -228,6 +235,77 @@ def test_saved_model_creation(self):
...
+    def prepare_config_and_inputs_for_common(self):
+        input_ids = tf.clip_by_value(
+            ids_tensor([self.batch_size, self.seq_length - 1], self.vocab_size),
+            clip_value_min=self.eos_token_id + 1,
+            clip_value_max=self.vocab_size + 1,
+        )
+        eos_tensor = tf.expand_dims(tf.constant([self.eos_token_id] * self.batch_size), 1)
+        input_ids = tf.concat([input_ids, eos_tensor], axis=1)
+
+    # TFBartForSequenceClassification does not support inputs_embeds
```
Comment (sgugger) [inline tests/models/bart/test_modeling_tf_bart.py:247]: There is no reason to skip this test for TFBartForSequenceClassification. Why doesn't it support inputs_embeds when the base model does?

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (sgugger) [reply tests/test_modeling_tf_common.py:173]: There shouldn't be any modification in the common tests for a PR adding a new model.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. There is some additional work to do in the tests, I think.

File: tests/models/bart/test_modeling_tf_bart.py (hunk 3 above)
Comment (ydshieh) [reply tests/models/bart/test_modeling_tf_bart.py:247]: I also asked myself, but found the same test is skipped in Bart testing file for `BartForSequenceClassification`. There is probably some good reason, but it has not been specified at that time.

File: src/transformers/models/bart/modeling_tf_bart.py (hunk 1 above)
Comment (uglyboxer) [reply src/transformers/models/bart/modeling_tf_bart.py]: Ah yes, you are correct.  Good catch. 

Removed.

File: tests/models/bart/test_modeling_tf_bart.py (hunk 3 above)
Comment (uglyboxer) [reply tests/models/bart/test_modeling_tf_bart.py:238]: As mentioned above in the clipping, comment, this model relies on access to the raw input_ids to correctly identify the embedding produced for the final `eos` token.   Happy to add a comment in this PR or fast follow with one if it helps.

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (uglyboxer) [reply tests/test_modeling_tf_common.py:173]: I agree this should go elsewhere, but I couldn't see a clear healthy path around `test_save_load_after_resize_token_embeddings`.  It programmatically creates `input_ids` that will be by definition incompatible with `TFBartForSequenceClassification`.

The base `input_ids` are compatible via the other updates in the PR, but a constant is applied to those ids in the test.   Open to options.

**Comment by uglyboxer** [comment]
@ydshieh @sgugger Thank you both for your insightful reviews.  Pushed some changes and posed a question back.

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (sgugger) [reply tests/test_modeling_tf_common.py:173]: If it's just to fix one test, you can overwrite it in the TFBart model tester.

File: tests/models/bart/test_modeling_tf_bart.py [hunk 4]
```diff
@@ -76,7 +77,11 @@ def __init__(
         self.bos_token_id = bos_token_id
 
     def prepare_config_and_inputs_for_common(self):
-        input_ids = ids_tensor([self.batch_size, self.seq_length - 1], self.vocab_size)
+        input_ids = tf.clip_by_value(
+            ids_tensor([self.batch_size, self.seq_length - 1], self.vocab_size),
+            clip_value_min=self.eos_token_id + 1,
+            clip_value_max=self.vocab_size + 1,
+        )
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply tests/models/bart/test_modeling_tf_bart.py:84]: You are right, there is also `clamp` in `BartModelTester.prepare_config_and_inputs`.
(I didn't realize this as the PT/TF model testing files don't have exactly test method names.)

File: tests/models/bart/test_modeling_tf_bart.py [hunk 5]
```diff
@@ -76,7 +77,11 @@ def __init__(
         self.bos_token_id = bos_token_id
 
     def prepare_config_and_inputs_for_common(self):
-        input_ids = ids_tensor([self.batch_size, self.seq_length - 1], self.vocab_size)
+        input_ids = tf.clip_by_value(
```
Comment (ydshieh) [inline tests/models/bart/test_modeling_tf_bart.py:80]: Maybe a short 1-line comment?

File: tests/models/bart/test_modeling_tf_bart.py [hunk 6]
```diff
@@ -76,7 +77,11 @@ def __init__(
...
+            ids_tensor([self.batch_size, self.seq_length - 1], self.vocab_size),
+            clip_value_min=self.eos_token_id + 1,
+            clip_value_max=self.vocab_size + 1,
+        )
         eos_tensor = tf.expand_dims(tf.constant([self.eos_token_id] * self.batch_size), 1)
```
Comment (ydshieh) [inline tests/models/bart/test_modeling_tf_bart.py:85]: 1-line comment like: add EOS tokens at the end of each sequence  ?

File: tests/models/bart/test_modeling_tf_bart.py (hunk 3 above)
Comment (ydshieh) [reply tests/models/bart/test_modeling_tf_bart.py:238]: Thanks, it's clear to me now. You are welcome to add a comment regarding the reason to skip, but no obligation ❤️

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:173]: @uglyboxer Isn't your updated version of `TFBartModelTester.prepare_config_and_inputs_for_common` already making the inputs compatible with `TFBartForSequenceClassification`?

As far as I know, we don't have a block corresponding to `if model_class.__name__ == "TFBartForSequenceClassification":` in `test_modeling_common.py` (for `pytorch` models). So I think this block here (added by you) is not necessary - or, at least, we can have the correct input preparation logic inside `test_modeling_tf_bart.py`.

Could you check again?

**Review by ydshieh — COMMENTED** [review]
Everything looks fine to me - except a final comment regarding the change in `_prepare_for_class`

**Comment by uglyboxer** [comment]
Thanks again @ydshieh

I'm afraid the special handling is necessary as the test `test_save_load_after_resize_token_embeddings`  does some extra magic to alter the input ids.  I took @sgugger 's suggestion to overwrite the test in the BartTester and move that logic into the test itself.    That should clear up common.

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 7]
```diff
@@ -1465,3 +1484,141 @@ def _reorder_cache(past, beam_idx):
...
+        loss = None if labels is None else self.hf_compute_loss(labels=labels, logits=logits)
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            return ((loss,) + output) if loss is not None else output
+
+        return TFSeq2SeqSequenceClassifierOutput(
+            loss=loss,
+            logits=logits,
+            cross_attentions=outputs.cross_attentions,
```
Comment (ydshieh) [inline src/transformers/models/bart/modeling_tf_bart.py:1597]: nit: Let's respect the ordering, i.e. put this after `decoder_attentions` :-)

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 8]
```diff
@@ -1465,3 +1484,141 @@ def _reorder_cache(past, beam_idx):
...
+            decoder_hidden_states=outputs.decoder_hidden_states,
+            decoder_attentions=outputs.decoder_attentions,
+            encoder_last_hidden_state=outputs.encoder_last_hidden_state,
+            encoder_hidden_states=outputs.encoder_hidden_states,
+            encoder_attentions=outputs.encoder_attentions,
+        )
+
+    def serving_output(self, output):
+        logits = tf.convert_to_tensor(output.logits)
+        cross_attns = tf.convert_to_tensor(output.cross_attentions) if self.config.output_attentions else None
```
Comment (ydshieh) [inline src/transformers/models/bart/modeling_tf_bart.py:1608]: ordering

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 9]
```diff
@@ -1465,3 +1484,141 @@ def _reorder_cache(past, beam_idx):
...
+        pkv = tf.tuple(output.past_key_values)[1] if self.config.use_cache else None
+        dec_hs = tf.convert_to_tensor(output.decoder_hidden_states) if self.config.output_hidden_states else None
+        dec_attns = tf.convert_to_tensor(output.decoder_attentions) if self.config.output_attentions else None
+        enc_hs = tf.convert_to_tensor(output.encoder_hidden_states) if self.config.output_hidden_states else None
+        enc_attns = tf.convert_to_tensor(output.encoder_attentions) if self.config.output_attentions else None
+
+        return TFSeq2SeqSequenceClassifierOutput(
+            logits=logits,
+            cross_attentions=cross_attns,
```
Comment (ydshieh) [inline src/transformers/models/bart/modeling_tf_bart.py:1617]: ordering

File: src/transformers/models/bart/modeling_tf_bart.py [hunk 10]
```diff
@@ -1465,3 +1484,141 @@ def _reorder_cache(past, beam_idx):
                 tuple(tf.gather(past_state, beam_idx, axis=0) for past_state in layer_past[:2]) + layer_past[2:],
             )
         return reordered_past
+
+
+@add_start_docstrings(
+    """
+        Bart model with a sequence classification/head on top (a linear layer on top of the pooled output) e.g. for
+        GLUE tasks.
```
Comment (ydshieh) [inline src/transformers/models/bart/modeling_tf_bart.py:1492]:
```suggestion
    Bart model with a sequence classification/head on top (a linear layer on top of the pooled output) e.g. for
    GLUE tasks.
```

**Review by ydshieh — APPROVED** [review]
Thank you @uglyboxer Level a few nits, but we are ready to merge!

cc @sgugger  for final approval.

**Review by sgugger — APPROVED** [review]
Good to go with the nits, thanks for bearing with us!

**Comment by uglyboxer** [comment]
> Good to go with the nits, thanks for bearing with us!

Happy to take care of them!  This is my first PR, thanks for all the help, and the seamless process.

**Comment by ydshieh** [comment]
> > Good to go with the nits, thanks for bearing with us!
> 
> Happy to take care of them! This is my first PR, thanks for all the help, and the seamless process.

You are doing a great job! 💯
